### ISSUE TITLE
Playing with ADAM

### ISSUE THREAD

**Issue by snakers4** [SeanNaren/deepspeech.pytorch#407]
Hi everyone!

Many thanks for an amazing repo!

So far I could achieve stable / parallelized training on a modified version of the code in this repo:
- stft instead of MFCC;
- librosa / numpy / scipy augs instead of sox;
- CNN based models instead of LSTM;
- Russian;

But I noticed that you use SGD and everyone seems to use SGD in this field.
I am planning to run a few experiments with ADAM, does anyone have any tips / experience in this?

So far from what I know from NLP / CV:
- ADAM likes lrs around 1e-3 / 1e-4, i.e. usually lower than SGD;
- ADAM is more aggressive, usually boosts training speed 2x-10x;
- Converges a bit worse, 2-3pp usually;

**Comment by tugstugi** [comment]
@snakers4
I usually use Adam with the Noam scheduling. But for training a speech recognition with the CTC loss, choosing a hyper parameter for Adam seems to be really difficult. Adam has a much narrower LR range where it converges somehow OK, in other hand SGD seems to be more robust for picking a hyper parameter. What really helps for training faster, is precomputing the spectrograms and do data augmentation only on the spectograms.

**Comment by snakers4** [comment]
I see.
We solved the fast pre-computation issue even for huge datasets, so I guess this issue is pointless then.

**Comment by Marcovaldong** [comment]
However, Adam can decrease the loss to a samll value in very short time, SGD needs more time to get the same performance. Can I use Adam firstly, when the loss is small enough, then I change to use SGD? @tugstugi @snakers4

**Comment by buriy** [comment]
@Marcovaldong i'm afraid that's only valid for images

**Comment by Marcovaldong** [comment]
@buriy I am training CTC and RNN-Transducer. I tried SGD and Adam, I found SGD is slower than Adam.
However, with adam,  I cannot get the performance which my colleague got using SGD. 

Why, can you tell me, I'll be very appreciate.

**Comment by buriy** [comment]
@Marcovaldong this is typical for Adam to converge faster but get lower final accuracy, and if you would try to tune the hyperparameters for better accuracy (reduce beta, or maybe add more L2 regularization) -- the speedup disappears. That's similar to training SGD with higher LR ;)
